python/lsst/cp/pipe/linearity.py

```diff
@@ -0,0 +1,235 @@
+            ampName = amp.getName()
+            if (len(inputPtc.visitMask[ampName]) == 0):
+                self.log.warn(f"Mask not found for {ampName} in non-linearity fit. Using all points.")
+                mask = np.repeat(True, len(inputPtc.rawExpTimes[ampName]))
+            else:
+                mask = inputPtc.visitMask[ampName]
+
+            timeVector = np.array(inputPtc.rawExpTimes[ampName])[mask]
+            meanVector = np.array(inputPtc.rawMeans[ampName])[mask]
+
+            if self.config.linearityType in ['Polynomial', 'Squared', 'LookupTable']:
+                polyFit = np.zeros(self.config.polynomialOrder + 1)
+                polyFit[1] = 1.0
+                polyFit, polyFitErr, chiSq = fitLeastSq(polyFit, timeVector, meanVector, funcPolynomial)
+
+                # Truncate the polynomial fit
+                k1 = polyFit[1]
+                linearityFit = [-coeff/(k1**order) for order, coeff in enumerate(polyFit)]
+                significant = np.where(np.abs(linearityFit) > 1e-10, True, False)
+                self.log.info(f"Significant polynomial fits: {significant}")
+
+                if self.config.linearityType == 'Squared':
+                    linearityFit = [linearityFit[2]]
+                elif self.config.linearityType == 'LookupTable':
+                    # Use linear part to get time at wich signal is maxAduForLookupTableLinearizer DN
+                    tMax = (self.config.maxLookupTableAdu - polyFit[0])/polyFit[1]
+                    timeRange = np.linspace(0, tMax, self.config.maxLookupTableAdu)
+                    signalIdeal = polyFit[0] + polyFit[1]*timeRange
+                    signalUncorrected = funcPolynomial(polyFit, timeRange)
+                    lookupTableRow = signalIdeal - signalUncorrected  # LinearizerLookupTable has corrections
+
+                    linearizer.tableData[tableIndex, :] = lookupTableRow
+                    linearityFit = [tableIndex, 0]
+                    tableIndex += 1
+            else:
+                polyFit = [0.0]
+                polyFitErr = [0.0]
+                chiSq = np.nan
+                linearityFit = [0.0]
+
+            linearizer.linearityType[ampName] = 'self.config.linearityType'
+            linearizer.linearityCoeffs[ampName] = linearityFit
+            linearizer.linearityBBox[ampName] = amp.getBBox()
+            linearizer.fitParams[ampName] = polyFit
+            linearizer.fitParamsErr[ampName] = polyFitErr
+            linearizer.fitChiSq[ampName] = chiSq
+            self.debugFit('solution', timeVector, meanVector, linearizer, ampName)
+
+        linearizer.validate()
+        linearizer.updateMetadata(setDate=True)
+        provenance = IsrProvenance(calibType='linearizer')
+
+        return pipeBase.Struct(
+            outputLinearizer=linearizer,
+            outputProvenance=provenance,
+        )
+
+    def debugFit(self, stepname, timeVector, meanVector, linearizer, ampName):
+        """
```

Comment:
No docstring here?  I guess un in L197 you are calling this function to produce linearity plots for debugging while the code is running, right? (like in the crosstalk code)

Reply:
That's the plan.  Since this can be run independent of the previous steps (ISR, PTC), the results can be checked through the debug code.  I've added the missing docstring.